### ISSUE TITLE
SSL Error: Certificate chain is invalid.

### ISSUE THREAD

**Issue by ganySA** [anurodhp/Monal#406]
Hi

I have setup a standalone ejabberd server on my localhost when i try connect to it i can see the connection coming through but i get an error that the certificate chain is invalid.

I have tried to turn off all the TLS and other functionality in monal but that does not seem to help.

Any ideas?

Thanks
___________________________________
[ ] I have cross-check this overview https://github.com/anurodhp/Monal/issues/322 as well es filtered for regarding labels https://github.com/anurodhp/Monal/labels

**Comment by m2x4m4l42n** [comment]
I had the same issue with my server, which is using self-signed certificates. Toggling off "validate certificate" did the trick for me but it seems to take a few moments to actually connect.

**Comment by anurodhp** [comment]
tried it with a self signed cert, it does work.

**Comment by keenonkites** [comment]
I'm running 5.0.1 (761) on Big Sur.... getting the error about invalid certificate chain.... where can I disable the validation of the certificate ?
Thanks for helping

**Comment by FriedrichAltheide** [comment]
https://github.com/monal-im/Monal/wiki/FAQ---Frequently-Asked-Questions#why-doesnt-monal-allow-self-signed-or-expired-certificates

**Comment by keenonkites** [comment]
Ok thanks, unfortunately that doesn't help me. We're running prosody on an internal server with no connection from the outside world and the servername is something.local .... so we would not run into security problems here. And I do not think we can get an official certificate for a hostname in the .local top level domain.
If there's no chance to tweek that (under the hood somehow) I've to look for a different client unfortunately.

**Comment by sirrosh** [comment]
> https://github.com/monal-im/Monal/wiki/FAQ---Frequently-Asked-Questions#why-doesnt-monal-allow-self-signed-or-expired-certificates

For me it sounds like blithering idiocy and renders the app completely unusable in a secure corporate environment...  but whatever.

This, however, does not explain why monal does not trust manually installed root CAs on ios. Could you please provide an answer to this?

BTW, Conversations.im does not pretend being smarter than its user, you can adapt its trust policy to your needs.

**Comment by tmolitor-stud-tu** [comment]
@sirrosh First of all: We develop Monal in our free time and in the last two years we are working on this project no company has ever reached out to us (and I doubt that ever happened).

You are talking about a custom CA, the wiki article is about **self-signed certs** those are two **distinct** things. Disabling certificate checks altogether is **really dangerous** and no, we won't allow users to be “smart” and shoot themselves in their foot.

That said, a correctly created and imported CA works, in fact some Monal users already did this and reported their success.

As for your question: a quick google search pointed me to this answer: <https://support.apple.com/en-us/HT204477>

If it still doesn't work, you are more than welcome to provide your OpenSSL config file, that was used to create the CA, so that we can give it a try. Please check that your certificate and server config is following Apples recommendations and guidelines (>= TLS 1.2, no IP-address in the certificate's common name etc.).

**In the future, please be so kind to open a new ticket if your topic differs or the ticket is closed for a long time. That way it is easier for all of us to follow the conversations. It's always possible to reference an older ticket in a new one, if you think that they might relate.**

And if you like Conversations better, there is always the possibility to switch to Android devices instead. If you do, please be so kind to consider donating a few bucks to Daniel (the Conversations maintainer) for his remarkable work.

**Comment by sirrosh** [comment]
@tmolitor-stud-tu, thank you for your time and for a dozen of advices. Now I feel sorry for waisting your free time...
And of course our company is buying every instance of Conversations, a great work never passes unnoticed.

**Comment by shodanx2** [comment]
Hello,

I would like to contribute user experience we had with this software

My xmpp server is running ejabberd with a self-signed certificate

It works with my friend, but then I sent my apple-user friend his xmpp login and password

He downloaded monal

When he entered his credentials, the following unclear error message appear

![image](https://user-images.githubusercontent.com/10621885/177060303-c4f04aec-8884-4832-ae10-862af893f8eb.png)

I say unclear because, the error message does not mention if the self-signed is the cause or if there is another problem. I can only assume that this is because it is a self-signed certificate.

I sent my friend a zip of all the *.pem files and told him he has to add them to his phone's keyring somehow

He responded, this is too complicated let's just use facebook instead and then uninstalled monal

So, here is a mockup of how change the user interface to result in a positive user experience

![image](https://user-images.githubusercontent.com/10621885/177061101-bc511667-c52b-47ad-bf4b-7f52615a013f.png)

I imagine this is the experience of 99% of XMPP users

**Comment by licaon-kter** [comment]
Certificates are free since 2016 or so, any reason not to use a non-self-signed?

**Comment by shodanx2** [comment]
third party dependence, internet dependence and extra complexity
If you're making baby's first xmpp server, your server will have self-signed certificates
Whatever the reason why the server presents self-signed certificate

The solution is to ask the user "hey, I can't confirm this server's certificates because there are no other reocrds available, the signature is [insert bunch of numbers here], do you want to use it anyway ?"  and then [x] trust this certificate forever

That's how gajim and conversation do it and it works (gajim url preview still breaks, you have to also disable HTTPS verification in the url preview options)

**Comment by licaon-kter** [comment]
Complications on top of complications. Ok, the server works with other clients? Get a proper cert and setup cron to update it. 

Independence? So you've setup accounts on your .onion domain too? :)

**Comment by tmolitor-stud-tu** [comment]
I can cite the wiki article linked above, in case you missed it:

Self-signed certificates are an anachronism.

Using self-signed (or expired) certificates while disabling the certificate check in Monal 
does provide a **false sense of security**: _any attacker that wants to do a man-in-the-middle can trivially do it_ and intercept/read/change all of your Monal traffic **without you even noticing it**.

Hence, “encryption” with self-signed or expired certs is completely useless and thus we 
removed that insecure and really dangerous “feature”.

**Solution:** Create a real not self-signed certificate for your server (Let's Encrypt and 
many others provide free certificates not costing a cent).  
See [Let's Encrypt: getting started](https://letsencrypt.org/getting-started/)

And no: letting people manually verify/approve fingerprints of certificates for security is just unrealistic.

**Comment by sirrosh** [comment]
> Hence, “encryption” with self-signed or expired certs is completely useless and thus we removed that insecure and really dangerous “feature”.

I find it rather cute when somebody decides something for me. And I always wonder what makes one think he's smarter than someone else ;)
 
> And no: letting people manually verify/approve fingerprints of certificates for security is just unrealistic.

Why?

**Comment by anurodhp** [comment]
Development time is a limited resource and work has to be prioritized. This is the nature of all software projects but doubly so when the labor is free. This feature made sense when certs cost money. That isn’t the case anymore and it is far simpler to just get a free valid cert. 
There may be a use case for this feature but it is niche and it’s not something that is a priority for developers. People decide things for you because volunteers decide how to use their own  free time. If you feel strongly, please submit a pull request.

**Comment by FriedrichAltheide** [comment]
To clarify: we will not change the current behaviour and we are not going to accept PRs that will add a "trust any cert" option or anything similar. @tmolitor-stud-tu and I decided  that we will only allow certs signed by a trusted CA. Either use any of the already trusted CAs to sign your cert **or** create your own CA (has to conform to apples guidelines) sign your cert and import it into the device.

**Comment by shodanx2** [comment]
Hello,

@licaon-kter, 
yes it works with gajim and conversation for several months now, both asked the user to accept the self-signed cert on the first login, gajim needed to have "https verification" disable in the url preview setting for the url preview to work

@tmolitor-stud-tu,
"Self-signed certificates are an anachronism" Is needlessly inflamatory 
Accepting a self-signed certificate opens you to a MITM attack only on the first connection when you accept the certificate. 
After that you would detect a MITM certificate. (although the FAQ is responding to just disabling all certificate checking, which is not what I'm asking here)

@sirrosh,
I do not doubt this was said with good intentions

@anurodhp
I agree, as I said in the first line of my first message, I am just reporting our user experience and why Monal didn't work for us.
There is a reason why facebook is on everyone's phone and xmpp is not.

@FriedrichAltheide 
I presume the client could just download the certification files and add them to the device keyring with user approval. Asking the user to find the certificate files, format them in apple approved format, copy them to the device and then figure out how to add them is a major roadblock that me and my friend failed to clear.   And we're both 40 year olds who have worked in IT all our lives. I think it's not realistic to expect grandma to figure this out


------

Resolution

I sent all of the server's 4 *.PEM files through facebook to my friend. He responded 

"Monal still does not work after I installed the PEM files, I installed Siskin and it worked on the first try"

So for us the problem is solved.
